Modify CRIU to dump + restore thread cgroups

angieenee opened the pull request:
Currently CRIU dump and restores only membership of processes in cgroups. Any information about membership of threads is lost, e.g. if a process is in cgroup “c1” and its thread is in cgroup “c2”, after C/R both will be in cgroup “c1” and cgroup “c2” won’t even exist on the restore machine. The objective is to fix that behaviour so that threads will be in the same cgroups after restore as they were before checkpoint.

Signed-off-by: Angie Ni <avtni@google.com>

Commit:
Modified parasite to dump thread cgroup

Dump proc/self-thread/cgroup instead of proc/self/cgroup by calling
compel_run_in_thread with PARASITE_CMD_DUMP_CGROUP. After succeeding,
log and verify the contents of the cgroup file.

Added new field in ThreadCoreEntry proto to store cgroup ID + modified
dump_task_cgroup to take in pid as input instead of pstree_item. Logs
identify that correct cgroups have been dumped.

Signed-off-by: Angie Ni <avtni@google.com>

angieenee commented on test/zdtm/static/cgroup_threads.c:
```diff
@@ -69,7 +69,9 @@ static int cg_check(char *name)
 	fclose(cgf);
 
-	return found ? 0 : -1;
+	// return found ? 0 : -1;
```
I changed the test, so please ignore this for now.

angieenee commented on criu/parasite-syscall.c:
```diff
@@ -165,32 +165,46 @@ int parasite_dump_thread_seized(struct parasite_thread_ctl *tctl,
+		goto err_rth;
+	}
+
+	thread_cgroup = cgargs->contents;
```
Use memcmp to check

angieenee commented on criu/cgroup.c:
```diff
@@ -716,32 +716,31 @@ static int collect_cgroups(struct list_head *ctls)
 		return -1;
 
-	cs = get_cg_set(&ctls, n_ctls, item);
+	cs = get_cg_set(&ctls, n_ctls, pid);
```
The third argument requested by `get_cg_set` is a bool that is only used once in `if (collect && collect_cgroups(&cs->ctls))` -- since an item that is NULL would have a pid that is also NULL, replacing item with pid will not change the outcome unless future changes are made.

angieenee commented on criu/cgroup.c:
```diff
@@ -716,32 +716,31 @@ static int collect_cgroups(struct list_head *ctls)
 }
 
-int dump_task_cgroup(struct pstree_item *item, u32 *cg_id, struct parasite_dump_cgroup_args *args)
+int dump_task_cgroup(struct pid *id, u32 *cg_id, struct parasite_dump_cgroup_args *args)
```
This function previously passed in `pstree_item` to check if it was NULL. If the item is NULL, then its pid should also be NULL. Unless future changes are implemented, it is sufficient to pass in its pid instead.

mclapinski replied:
not a big fan of the name `id`

kashyapp commented on criu/cr-dump.c:
```diff
@@ -846,8 +847,10 @@ static int dump_task_thread(struct parasite_ctl *parasite_ctl,
 				const struct pstree_item *item, int id)
 {
 	struct parasite_thread_ctl *tctl = dmpi(item)->thread_ctls[id];
+	struct parasite_dump_cgroup_args cgroup_args, *info = NULL;
```
I haven't understood why the info pointer is needed. Even in the existing code. Why can't we just use &cgroup_args in place of info.

angieenee replied:
For both the existing + added code, the distinction is what we're doing with that memory. 

For example, [the existing code calls &cgroup_args directly](https://github.com/checkpoint-restore/criu/blob/958b750a3d6f3bbebfe8cbc800952dc39587485e/criu/cr-dump.c#L779) when the parasite is involved. As the name `info` imples, the existing code calls info when non-parasitic code needs the contents of the cgroup_args struct. 

I believe it's a matter of manipulation vs. collection, and for consistency-sake, I followed how the existing code achieved this. I wouldn't have been able to figure this out without some of your *pointers* -- I'd be happy to explore this more if we need more clarification!

mclapinski replied:
I agree with @kashyapp

mclapinski replied:
please remove the `info` variable

angieenee commented on criu/cr-dump.c:
```diff
@@ -1018,8 +1030,11 @@ static int dump_task_threads(struct parasite_ctl *parasite_ctl,
 		}
-		if (dump_task_thread(parasite_ctl, item, i))
+		if (dump_task_thread(parasite_ctl, item, i)) {
+			pr_info("avtni@ failed to dump task thread\n");
```
Convert this to `pr_err` instead?

mclapinski commented on criu/cr-dump.c:
```diff
@@ -1018,8 +1030,11 @@ static int dump_task_threads(struct parasite_ctl *parasite_ctl,
+			pr_info("avtni@ failed to dump task thread\n");
 			return -1;
+		}
+			// avtni@ return -1;
```
please revert this one line

mclapinski reviewed (approved):
Just some minor comments. Besides that, really nice.

mclapinski reviewed (changes requested):
but fix the cgroup02 test first

angieenee commented:
Oh dear, time to squash.

angieenee commented on criu/cgroup.c:
```diff
@@ -716,32 +716,31 @@ static int collect_cgroups(struct list_head *ctls)
 		return -1;
 
-	cs = get_cg_set(&ctls, n_ctls, item);
+	cs = get_cg_set(&ctls, n_ctls, item_pid);
```
`item_pid` was previously local variable `pid`, which caused the cgroups02 test to fail -- `pid` would prevent the `get_cg_set` function from fully dumping the cgroups.

angieenee commented on test/zdtm/static/link_remap.9:
```diff
@@ -0,0 +1 @@
+000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 ...
```
0n n0, Angie, please deal with this.

mclapinski reviewed (changes requested):
1. Remove the useless file (it's causing test issues).
2. `rebase -i` to get rid of the cgroup_threads test commit.

mclapinski commented on criu/cr-dump.c:
```diff
@@ -846,8 +847,10 @@ static int dump_task_thread(struct parasite_ctl *parasite_ctl,
+	struct parasite_dump_cgroup_args cgroup_args;
 	struct pid *tid = &item->threads[id];
 	CoreEntry *core = item->core[id];
+	ThreadCoreEntry *thread_core = core->thread_core;
```
since you only use `thread_core` in one place, I'd remove the variable and use `core->thread_core` directly.

mclapinski commented on test/zdtm/static/cgroup_threads.c:
```diff
@@ -70,6 +70,7 @@ static int cg_check(char *name)
 	fclose(cgf);
 
 	return found ? 0 : -1;
+
```
revert

Commit:
thread compatible param for prepare_task_cgroup

Replaced pstree_item argument to the current pstree_item's cg_set and
its parent pstree_item's cg_set. Moved zombie parent check outside of
prepare_task_cgroup and into caller instead.

Signed-off-by: Angie Ni <avtni@google.com>

kashyapp commented on test/zdtm/static/cgroup_threads.c:
```diff
@@ -8,6 +8,11 @@
 #include <pthread.h>
 #include "zdtmtst.h"
 
+#ifdef __NR_seccomp
```
Drop this block

kashyapp commented on test/zdtm/static/cgroup_threads.c:
```diff
@@ -17,6 +22,8 @@ static const char *cgname = "zdtmtst";
 #define SUBNAME	"subcg_threads"
 #define SUBNAME2 SUBNAME"/subsubcg"
 
+#ifdef __NR_seccomp
```
The test for __NR_seccomp is to see if some specific support (syscall?) exists before enabling the test. Our test doesn't depend on that syscall being present. Instead you could 

#define BROKEN

#ifdef BROKEN

and invert the if/else blocks

kashyapp commented on criu/cgroup.c:
```diff
@@ -1208,32 +1207,27 @@ static int move_in_cgroup(CgSetEntry *se, bool setup_cgns)
 }
 
-int prepare_task_cgroup(struct pstree_item *me)
+int prepare_task_cgroup(u32 cg_set, u32 parent_cg_set)
```
Instead of pushing code out of this into its parent, why not extract out another function, say `prepare_cgroup` which is called by `prepare_task_cgroup`.

With the changes as they are now, the calling code in `restore_task_in_children` is getting polluted.

Anyway, this is not important now, come back to it when you have end-to-end working implementation.

kashyapp commented on criu/cr-restore.c:
```diff
@@ -1781,10 +1781,13 @@ static int create_children_and_session(void)
 {
 	struct cr_clone_arg *ca = _arg;
+	struct pstree_item *parent;
+	u32 parent_cg_set;
```
Why is this not initialized? To `0` perhaps.

angieenee replied:
Good catch! Thank you.

kashyapp commented on criu/cr-restore.c:
```diff
@@ -1845,18 +1848,26 @@ static int restore_task_with_children(void *_arg)
 		goto err;
 
+	/* Zombies and helpers can have cg_set == 0 so we skip them */
+	while (parent && !rsti(parent)->cg_set && parent->parent) {
```
The original code didn't have the `&& parent->parent` check in the if condition. Why is it needed here?

angieenee replied:
If parent->parent (the grandparent item, if you will) is NULL, we may get another segfault for calling `rsti(parent)->cg_set` after assigning parent->parent as the new parent.

kashyapp replied:
But it is functionally different from the existing code. Take this chain for example
a(cg_set = 0) <- b (cg_set = 0) <- c (cg_set=0) <- current
Your loop will exit with parent = a

But the original code would have exited the loop with 
parent = null

This makes the code materially different.

kashyapp replied:
Also, grandparent being NULL is valid. The problem is that we've changed the logic. In the original code, if parent is null, it takes root_cg_set. But with the example I just gave, it will take cg_set = 0.

angieenee replied:
My intention with this change is so `rsti(parent)->cg_set = 0` can make the following check in `prepare_task_cgroups` work similarly to how the original code intended:
```
if (parent_cg_set)
		current_cgset = parent_cg_set;
else
		current_cgset = root_cg_set;
```
I see your point in how the added `parent->parent` condition may exclude a NULL grandparent, and I will modify the code so it does not exclude the NULL grandparent -- however, I believe the code as is would exit the loop with parent = a, pass in parent_cg_set = 0 to `prepare_task_cgroups` and still assign current_cgset = root_cg_set.

mclapinski replied:
Although I think Angie is right and it would work fine, I agree with Kashyap that it would be easier/clearer if you just moved the whole initialization of `current_cgset` here.

angieenee commented on criu/cgroup.c:
```diff
@@ -1242,7 +1237,9 @@ int prepare_task_cgroup(struct pstree_item *me)
-	return move_in_cgroup(se, !me->parent);
+	pr_info("avtni@ current_cgset: %d\n", current_cgset);
+
+	return move_in_cgroup(se, !parent_cg_set);
```
pass in a bool argument to represent if parent is NULL / current is root

mclapinski replied:
It's hard to understand if the behavior here got changed or not. It would be clearer if you just passed another argument like `is_root_task`.

angieenee commented:
I've written buggy code for the final step (14 failed test cases 😔), but I'll push those changes _after_ approval for Steps 1-4.

mclapinski commented on criu/cr-restore.c:
```diff
@@ -1850,13 +1853,22 @@ static int restore_task_with_children(void *_arg)
+	/* Zombies and helpers can have cg_set == 0 so we skip them */
+	while (parent && !rsti(parent)->cg_set) {
+		parent = parent->parent;
+		current_cg_set = rsti(parent)->cg_set;
```
possible segfault, just move the code from `prepare_task_cgroup` here

mclapinski commented on criu/cgroup.c:
```diff
@@ -1242,8 +1230,7 @@ int prepare_task_cgroup(struct pstree_item *me)
 	 */
-
-	return move_in_cgroup(se, !me->parent);
+	return move_in_cgroup(se, !is_root_task);
```
wrong, remove the `!`

mclapinski commented on criu/cr-restore.c:
```diff
@@ -1850,13 +1853,22 @@ static int restore_task_with_children(void *_arg)
 	 * just have it inherited.
 	 */
-	if (prepare_task_cgroup(current) < 0)
+	if (prepare_task_cgroup(rsti(current)->cg_set, current_cg_set, parent) < 0)
```
passing `parent` here is wrong

mclapinski commented on criu/cr-restore.c:
```diff
@@ -1781,10 +1781,13 @@ static int create_children_and_session(void)
 	int ret;
 
 	current = ca->item;
+	parent = current->parent;
```
possibly remove this line and move the assignment to the declaration line (don't use current then)
